### PR TITLE
Initial implementation of aggregation job creator.

### PR DESCRIPTION
**Pull request by BranLwyd**
The concurrency part was the most challenging piece: I would have liked
to have the task-creation worker (i.e. AggregationJobCreator::run()) be
"tightly linked" to the job-creation workers (i.e.
AggregationJobCreator::run_for_task()) in the sense that a
task-discovery worker shutting down would be noticed & cleaned up
immediately (in a select loop). Sadly, the lifetime/ownership issues
involved made this very challenging. Instead, I settled for a strategy
where the task-discovery worker owns a oneshot sender for each
job-creation worker, and is responsible to notice when a task is no
longer relevant; actual shutdown of the job-creation worker is loosely
coupled and may happen later. This approach is hypothetically racy and
definitely slower to notice a lingering no-longer-existent task, but I
think those issues are marginal enough that this is an acceptable
strategy -- I think the worst outcome we'll see is, if we delete a task,
we'll see its job-creation worker produce errors until the next time the
task-creation worker updates its view of existing tasks.

### CODE REVIEW

**Comment by BranLwyd** [comment]
Closes #117.

**Comment by BranLwyd** [comment]
One question: is there any way currently to know/estimate the size of the messages we'll be encoding? I ask because aggregate requests impose a limit of 64KiB of preparation message data, which we could bump into quite easily if the messages become large. Right now we'd just have to guess at a lowest-common-denominator number that would allow our aggregation jobs to not break the limit for even the VDAF with the largest messages.

**Comment by divergentdave** [comment]
I did the math on Prio3 prepare messages, `Prio3PrepareMessage` encodes `VERIFIER_LEN` many field elements, optionally followed by a PRG seed if joint randomness is used. `Transition` encodes a nonce, a one-byte discriminant, a two-byte length field, and then the prepare message as an `opaque`.

### Prio3Aes128Count

bytes per field element: 8
verifier_len: 4
joint randomness seed?: no
seed size: N/A, but would be 16 bytes
Prio3PrepareMessage encoded size: 4 * 8 + 0 = 32 bytes
Transition size: 8 + 8 + 1 + 2 + 32 = 51 bytes

### Prio3Aes128Sum
bytes per field element: 16
verifier_len: 3
joint randomness seed?: yes
seed size: 16 bytes
Prio3PrepareMessage encoded size: 3 * 16 + 16 = 64 bytes
Transition size: 8 + 8 + 1 + 2 + 64 = 83 bytes

### Prio3Aes128Histogram
bytes per field element: 16
verifier_len: 3
joint randomness seed?: yes
seed size: 16 bytes
Prio3PrepareMessage encoded size: 3 * 16 + 16 = 64 bytes
Transition size: 8 + 8 + 1 + 2 + 64 = 83 bytes

For these three instantiations, we can go up to 789 reports per aggregation job before we get into trouble. Of course, there's no generic guarantees about encoded message size at the VDAF level. It's possible a VDAF may have variable-length prepare messages as well, though that seems like a stretch in this information-theoretic context, plus one hopes future VDAFs would try to put reasonable bounds on their communication requirements.

**Comment by BranLwyd** [comment]
Thanks, this is helpful in choosing appropriate batch sizes for our initial pilot. I suggest that we eventually want this kind of estimation in code, since as you note there are no generic guarantees about message size across all VDAFs. (And I had a similar thought to you about variable message sizes; I guess as long as we can get a reasonable upper bound, we might be OK.)

This is especially painful since our current encoding implementation panics if we try to encode too much data into a variable-length field.

**Review by divergentdave — CHANGES_REQUESTED** [review]
Some thoughts for future follow-up:

- To optimize the efficiency of finding unaggregated client reports, we may want to consider denormalizing whether a report has ever been included in an aggregation job or not. If we add a boolean column for this to client_reports, update it when setting up a job, and include it in a clustered index, then each request for unaggregated reports becomes simpler for the database. Currently, we perform an index scan over all client reports for a task, and perform point lookups in the report aggregation table's index for each report. This may unnecessarily perform a subquery for each client report that has already been aggregated, but not deleted from the database. (assuming there aren't enough fresh reports to hit the LIMIT) With the denormalized boolean column and corresponding index, we would instead perform a single index range scan, and performance would be largely unaffected by retention of very old reports.
- We could add a facade for `tokio::time::Interval` to `Clock`, which would allow for better tests, without involving `tokio::time::pause()` and its side-effects.

File: test_util/src/lib.rs
Comment (BranLwyd) [reply test_util/src/lib.rs:5]: I would expect (without checking) that identical strings in .rodata would end up interned to a single string -- at least, that's how some other compilers work.

Still, switched to static.

Comment (BranLwyd) [reply test_util/src/lib.rs:5]: I would expect (without checking) that identical strings in .rodata would end up interned to a single string -- at least, that's how some other compilers work.

Still, switched to static.

File: janus_server/Cargo.toml
Comment (BranLwyd) [reply janus_server/Cargo.toml:45]: Removed.

File: janus_server/src/config.rs
Comment (BranLwyd) [reply janus_server/src/config.rs:65]: Done.

File: test_util/src/lib.rs
Comment (divergentdave) [reply test_util/src/lib.rs:5]: Ah, it appears you're right. I may be misremembering a conversation about unnecessary memcpys.

File: janus_server/src/datastore.rs
Comment (BranLwyd) [reply janus_server/src/datastore.rs:491]: Ah--late in the game, I decided to change the behavior to not include the nonce in the ordering requirement, but forgot to update the query (on the idea that we don't _need_ the ordering, and less constraints provide the DB more freedom to go fast). But I didn't update the comment :-). Updated comment to be accurate to what is implemented.

**Comment by BranLwyd** [comment]
With the latest commit I also updated the query used in `get_unaggregated_client_report_nonces_for_task`. I am following the pattern in https://explainextended.com/2009/09/16/not-in-vs-not-exists-vs-left-join-is-null-postgresql/ -- which is targeted at the (ancient) Postgres 8.4, but I would expect that Postgres would maintain efficient DB access patterns. Note this is unmeasured for now.

**Review by divergentdave — APPROVED** [review]
LGTM modulo merge conflicts

**Comment by BranLwyd** [comment]
(going to wait a little while before merging to let Tim review, if he wishes)

**Comment by tgeoghegan** [comment]
> (going to wait a little while before merging to let Tim review, if he wishes)

I do plan to review this today, but I can't promise I will get to it. If this change unblocks further work, you should go ahead and merge it. I can review this PR later and discuss any concerns I have with you or write them up as issues.

**Comment by BranLwyd** [comment]
> > (going to wait a little while before merging to let Tim review, if he wishes)
> 
> I do plan to review this today, but I can't promise I will get to it. If this change unblocks further work, you should go ahead and merge it. I can review this PR later and discuss any concerns I have with you or write them up as issues.

Merging as the `binary_utils` changes especially will be handy for the aggregation job driver. Still interested in review, I'll address any comments in a follow-on PR.

File: janus_server/src/bin/aggregation_job_creator.rs
Comment (tgeoghegan) [inline janus_server/src/bin/aggregation_job_creator.rs:242]: I think `panic!` here is the right thing to do because this most likely indicates a mismatch between the deployed DB schema and the schema version `aggregation_job_creator` was built against. However I think we should record a message with `error!` so that we get some indication that this is happening besides some pod getting stuck in a crash loop.

File: janus_server/Cargo.toml
Comment (BranLwyd) [reply janus_server/Cargo.toml:67]: I really want to introduce a "test-util" feature to our crate and use it instead of the hacky `#[doc(hidden)]` solution we use now, except that I think a crate can't control its own features when "self-importing" (e.g. for tests). :-(

Still, with how we're breaking the crates out, maybe we could do something like `#[cfg(test)]` OR `#[cfg(feature = "test-util")]`.

File: janus_server/src/bin/aggregation_job_creator.rs
Comment (BranLwyd) [reply janus_server/src/bin/aggregation_job_creator.rs:114]: (done, locally; I think I agree with adding `'static` to the `Clock` trait, but I'll leave the larger refactoring for now)

File: janus_server/src/bin/aggregation_job_creator.rs
Comment (BranLwyd) [reply janus_server/src/bin/aggregation_job_creator.rs:133]: It seems that the `Clock` bound is required (though I think a smart enough Rust compiler should be able to infer it? take that guess with a grain of salt, though):
```
error[E0277]: the trait bound `C: Clock` is not satisfied
   --> janus_server/src/bin/aggregation_job_creator.rs:130:9
    |
130 | impl<C> AggregationJobCreator<C> {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clock` is not implemented for `C`
    |
note: required by a bound in `AggregationJobCreator`
   --> janus_server/src/bin/aggregation_job_creator.rs:112:33
    |
112 | struct AggregationJobCreator<C: Clock + 'static> {
    |                                 ^^^^^ required by this bound in `AggregationJobCreator`
help: consider restricting type parameter `C`
    |
130 | impl<C: janus_server::time::Clock> AggregationJobCreator<C> {
    |       +++++++++++++++++++++++++++
```

OTOH, I'm not sure why the `'static` bound isn't required. Rust sometimes allows eliding lifetimes, maybe this is one case where the Rust compiler is smart enough to figure things out? I added `'static` anyway for consistency.

File: janus_server/src/bin/aggregation_job_creator.rs
Comment (BranLwyd) [reply janus_server/src/bin/aggregation_job_creator.rs:242]: Added an `error!`.

Note that DB schema skew isn't the only reason; I think we could also encounter this branch if we configured a VDAF that didn't yet have code support (like `poplar1`).

File: janus_server/src/bin/aggregation_job_creator.rs
Comment (BranLwyd) [reply janus_server/src/bin/aggregation_job_creator.rs:98]: The technically-correct-but-unhelpful answer is that it allows us to call `AggregationJobCreator::run()`, which receives `self` as an `Arc<Self>`.

Of course, this leads to the question of `run` takes an `Arc<Self>`. This is to allow `run` to spawn tasks (which run `AggregationJobCreator::run_for_task`), handing each task a reference to the `AggregationJobCreator` by cloning the `Arc` instead of the entire `AggregationJobCreator` -- we couldn't do this if `self` was a `&Self` due to lifetime issues, since the tasks run independently from `run`. This allows `run_for_task` and methods it calls to take a `&Self` rather than an `Arc<Self>`.

The other strategy, I suppose, would be to wrap "large" or otherwise unclonable pieces of the structure (e.g. the `Datastore`) in an `Arc`, rather than the entire structure. This is the strategy that `Aggregator` takes. But this would potentially lead to more reference-count juggling, and more importantly would mean we'd need to clone/copy each part of the `AggregationJobCreator` struct we needed to pass down to methods we call rather than passing down a reference to the entire `AggregationJobCreator` -- this would be painful since there are a number of configuration parameters, it's very nice to be able to pass them all along in a single parameter.

Why doesn't `Aggregator` use a similar strategy? Maybe it should, though it has fewer "configuration parameters" so for now managing an `Arc<Datastore>` is just as good.

File: janus_server/src/config.rs
Comment (BranLwyd) [reply janus_server/src/config.rs:80]: Hmm -- I sorta disagree here. To keep the same configuration semantics, I would need to have `Duration` (de)serialize from an integer number of seconds. (why seconds specifically, and not some other unit?) Then I would want the in-memory field to be named e.g. `tasks_update_frequency` while the equivalent YAML field would be named `tasks_update_frequency_secs` (since we want to have some context on what the numeric value means) -- not impossible, but more complexity.

Since this structure is supposed to mirror a configuration file, I think it is OK to temporarily use a number instead of a `Duration`. Note that we do translate to a `Duration` before using these values in program logic -- see https://github.com/divviup/janus/blob/2290fe4deb394d5b1532965fe045b4bca6a40300/janus_server/src/bin/aggregation_job_creator.rs#L101-L104.

Comment (tgeoghegan) [reply janus_server/src/config.rs:80]: I'm fine with that conclusion. It's not clear that spreading types whose purpose is encoding of network messages into config files is appropriate. Maybe what we want here is to move `message::{Time, Interval, Duration}` into `mod time` to decouple those types from the transport layer but I don't think that kind of code Feng Shui is urgent.

**Pull request merged**
